doc/xbgpu.design.b.rst

```diff
@@ -86,3 +86,35 @@ scaling to convert to a real value in :math:`(-0.5, 0.5)`. While this is
 still a deviation from uniformity, it does give a symmetric distribution.
 
 .. _curand: https://docs.nvidia.com/cuda/curand/index.html
+
+Data flow
+---------
+The host side of the beamforming is simpler than for correlation, because
+there is no accumulation. For simplicity, the output chunk size (in time) is
+set to the same as the input chunk size.
+
+Because the kernel operates on all beams together, there is only one instead
+of the :class:`.BPipeline` class, and it handles all the beams. This is in
+contrast to :class:`.XPipeline`, which only handles a single :class:`Output`.
+
+Unlike the correlation pipeline, the beamformer pipeline can be controlled
+dynamically by setting weights and gains. Transferring these to the GPU is
+somewhat expensive and should not be done for every chunk. Instead, there is
+an associated version number. When the weights/gains are updated, the master
+version number is incremented. Before processing a chunk on the GPU, this
+version number is compared to a version number associated with the GPU copy,
+and if they differ then an update is performed.
+
+Missing data handling
+^^^^^^^^^^^^^^^^^^^^^
+When some input data is not received, we wish to exclude it from the beam
+sums. At present this is handled by zeroing out missing data when it is
+received in :meth:`.XBEngine._receiver_loop`. This is potentially expensive if
+a lot of data goes missing. Should it prove too expensive, we could zero the
+data on the GPU, or pass metadata to the kernel to indicate which values
+should not be used.
+
+The presence of heaps is also tracked, to allow the ``beam_ants`` SPEAD item
+to be populated on transmission. This item is a compromise: ideally we'd like
```

Comment:
Apologies for being a non-zero pain here. One last comma-related flag.
> The presence of heaps is also tracked, to allow the ``beam_ants` SPEAD item to be populated on transmission.

I'd say either mention it in one breath or separate these points. I'd vote for two separate sentences.
> The presence of heaps is also tracked [great]. This is used to populate the ``beam_ants`` SPEAD item on transmission [also great].

Reply:
I deleted the comma. Tracking isn't valuable for itself - it needs a reason.